lib/libshare/smb.c

```diff
@@ -153,39 +157,75 @@ smb_retrieve_shares(void)
 					goto out;
 				}
 
-				strncpy(shares->name, name,
-					sizeof (shares->name));
-				shares->name [sizeof (shares->name) - 1] = '\0';
+				if (name != NULL) {
+					strncpy(shares->name, name,
+						sizeof (shares->name));
+					shares->name [sizeof (shares->name) - 1] = '\0';
+				}
 
-				strncpy(shares->path, path,
-				    sizeof (shares->path));
-				shares->path [sizeof (shares->path) - 1] = '\0';
+				if (path != NULL) {
+					strncpy(shares->path, path,
+						sizeof (shares->path));
+					shares->path [sizeof (shares->path) - 1] = '\0';
+				}
 
-				strncpy(shares->comment, comment,
-				    sizeof (shares->comment));
-				shares->comment[sizeof (shares->comment)-1] =
-				    '\0';
+				if (comment != NULL) {
+					strncpy(shares->comment, comment,
+						sizeof (shares->comment));
+					shares->comment[sizeof (shares->comment)-1] = '\0';
+				}
 
-				shares->guest_ok = atoi(guest_ok);
+				if (guest_ok != NULL) {
+					shares->guest_ok = atoi(guest_ok);
+				}
 
 				shares->next = new_shares;
 				new_shares = shares;
 
-				name = NULL;
-				path = NULL;
-				comment = NULL;
-				guest_ok = NULL;
+				if (path != NULL) {
+				    free(path);
```

Comment:
Calling `free()` with a NULL argument is OK and done elsewhere in the code.  It might make this simpler.